src/Features/Core/Portable/ConvertLinq/AbstractConvertLinqProvider.cs

```diff
@@ -0,0 +1,143 @@
+        {
+            protected readonly SemanticModel _semanticModel;
+            protected readonly CancellationToken _cancellationToken;
+
+            public AnalyzerBase(SemanticModel semanticModel, CancellationToken cancellationToken)
+            {
+                _semanticModel = semanticModel;
+                _cancellationToken = cancellationToken;
+            }
+
+            public async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+            {
+                var document = context.Document;
+                var cancellationToken = context.CancellationToken;
+                var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                var refactor = FindNodeToRefactor(root, context);
+                if (refactor == null)
+                {
+                    return;
+                }
+
+                if (refactor.ContainsDiagnostics)
+                {
+                    return;
+                }
+
+                if (TryConvert(refactor, out var result))
+                {
+                    if (!result.IsValid())
+                    {
+                        return;
+                    }
+
+                    context.RegisterRefactoring(new MyCodeAction(Title, c => UpdateDocumentAsync(root, document, result)));
+                }
+            }
+
+            protected abstract bool TryConvert(TRefactor refactor, out DocumentUpdate documentUpdate);
+
+            protected virtual TRefactor FindNodeToRefactor(SyntaxNode root, CodeRefactoringContext context) =>
+                root.FindNode(context.Span).FirstAncestorOrSelf<TRefactor>();
+
+            protected abstract string Title { get; }
+
+            private Task<Document> UpdateDocumentAsync(
+                SyntaxNode root,
+                Document document,
+                DocumentUpdate documentUpdate)
+            {
+                root = documentUpdate.UpdateRoot(root);
+                return Task.FromResult(document.WithSyntaxRoot(root));
+            }
+        }
+
+        protected sealed class DocumentUpdate
+        {
+            private ImmutableArray<(SyntaxNode Source, ImmutableArray<SyntaxNode> Destinations)> _updates;
+
+            public DocumentUpdate(SyntaxNode source, SyntaxNode destination)
+                => _updates = ImmutableArray.Create((source, ImmutableArray.Create(destination)));
+
+            public DocumentUpdate(SyntaxNode source, IEnumerable<SyntaxNode> destination)
+                => _updates = ImmutableArray.Create((source, ImmutableArray.CreateRange(destination)));
+
+            public bool IsValid()
+            {
+                return _updates.Any() && _updates.All(item => item.Source != null && item.Destinations != null);
+            }
+
+            public SyntaxNode UpdateRoot(SyntaxNode root)
+            {
+                foreach (var updateItem in _updates.OrderByDescending(update => update.Source.Span.End))
```

Comment:
needs comment.

Reply:
actually, updates of the same refactoring should not overlap. Therefore, the order is not required. Removed ordering.

---
In reply to: [175268106](https://github.com/dotnet/roslyn/pull/25362#discussion_r175268106) [](ancestors = 175268106)